server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -798,6 +799,24 @@ private ThreadContextStruct putResponseHeaders(Map<String, Set<String>> headers)
             return new ThreadContextStruct(requestHeaders, newResponseHeaders, transientHeaders, isSystemContext);
         }
 
+        private void logWarningHeaderThresholdExceeded(long threshold, Setting<?> thresholdSetting) {
+            // If available, log some selected headers to help identifying the source of the request.
+            // Note: Only Task.HEADERS_TO_COPY are guaranteed to be preserved at this point.
+            Map<String, String> selectedHeaders = new HashMap<>(requestHeaders);
+            selectedHeaders.keySet().retainAll(List.of(Task.X_OPAQUE_ID_HTTP_HEADER, Task.X_ELASTIC_PRODUCT_ORIGIN_HTTP_HEADER));
+
+            String messagePattern = HttpTransportSettings.SETTING_HTTP_MAX_WARNING_HEADER_SIZE.equals(thresholdSetting)
+                ? "Dropping a warning header{}, as their total size reached the maximum allowed of [{}] bytes set in [{}]!"
+                : "Dropping a warning header{}, as their total count reached the maximum allowed of [{}] set in [{}]!";
+
+            logger.warn(
+                messagePattern,
+                selectedHeaders.isEmpty() ? "" : Strings.format(" for request [%s]", selectedHeaders),
```

Comment:
Could we mention here that the request had no `X-Opaque-Id` header, to guide users towards setting this header for further investigation without needing to open a support case. Ideally also using `ReferenceDocs` to add a link to [the relevant docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/api-conventions.html#x-opaque-id) too.

Reply:
Great suggestion, thanks for the pointer @DaveCTurner